ast-utils.R

```diff
@@ -0,0 +1,139 @@
+is_function_used_in_var <- function(func_name, var_paths, var, main_code) {
+  var_paths_c <- var_paths
+  var_paths_c[[var]] <- NULL
+  any(
+    lapply(var_paths[[var]], 
+           function(x) is_function_used(func_name, var_paths_c, get_node(x, main_code), main_code)
+    ) == TRUE
+  )
+}
+
+is_function_used <- function(func_name, var_paths, sub_tree_root, main_code){
+    switch(expr_type(sub_tree_root),
+      constant = {
+        FALSE
+      },
+      symbol = {
+        var_paths_c <- var_paths
+        var_paths_c[[sub_tree_root]] <- NULL
+        any(
+          lapply(var_paths[[sub_tree_root]], 
+                 function(x) is_function_used(func_name, var_paths_c, get_node(x, main_code), main_code)
+          ) == TRUE
+        )
+      },
+      call = {
+        if (is_call(sub_tree_root, func_name)) {
+          return(TRUE)
+        } 
+        children <- as.list(sub_tree_root)
+        if(is_call(sub_tree_root, "<-")){
+          children[[2]] <- NULL
+        } 
+        any(
+          lapply(children, 
+                 function(x) is_function_used(func_name, var_paths, x, main_code)
+          ) == TRUE
+        )
+      },
+      expression = {
+          any(
+          lapply(as.list(sub_tree_root), 
+                 function(x) is_function_used(func_name, var_paths_c, x, main_code)
+          ) == TRUE
+        )
+      },
+      pairlist = {
+        any(
+          lapply(as.list(sub_tree_root), 
+                 function(x) is_function_used(func_name, var_paths_c, x, main_code)
+          ) == TRUE
+        )
+      },
+      stop("Don't know how to handle type ", typeof(sub_tree_root), call. = FALSE)
+    )
+}
+
+get_node <- function(path, code){
+  element <- code
+  for (direction in path){
+    element <- as.list(element)[[direction]]
+  }
+  return(element)
+}
+
+expr_type <- function(x) {
+  if (rlang::is_syntactic_literal(x)) {
+    "constant"
+  } else if (is.symbol(x)) {
+    "symbol"
+  } else if (is.call(x)) {
+    "call"
+  } else if (is.pairlist(x)) {
+    "pairlist"
+  } else if (is.expression(x)){
+    "expression"
+  } else {
+    typeof(x)
```

Comment:
Are you sure `typeof` could never return any of these other strings? If not, does R have some sort of user defined Enum type with which to replace them (or is there an enum of types built-in)?

Alternatively, use the string `"other"` instead of `typeof(x)` and match on `"other"` in the switch-statement.

Reply:
As I understand enums are not really available in R but I've seen people use named list as a replacement so I think the named list will do the job. The typeof(x) function is really usefull for debugging but for the final version your idea sounds like the better one. 
I will keep your comments in mind at a next PR.